library/Vanilla/Models/SSOModel.php

```diff
@@ -0,0 +1,319 @@
+     */
+    public function sso(SSOInfo $ssoInfo) {
+        $user = $this->getUser($ssoInfo);
+
+        if (!$user) {
+            // Allows registration without an email address.
+            $noEmail = $this->config->get('Garden.Registration.NoEmail', false);
+
+            // Specifies whether Emails are unique or not.
+            $emailUnique = !$noEmail && $this->config->get('Garden.Registration.EmailUnique', true);
+
+            // Allows SSO connections to link a VanillaUser to a ForeignUser.
+            $allowConnect = $this->config->get('Garden.Registration.AllowConnect', true);
+
+            // Will automatically try to link users using the provided Email address if the Provider is "Trusted".
+            $autoConnect = $ssoInfo['authenticatorIsTrusted']
+                && $allowConnect
+                && $emailUnique
+                && $this->config->get('Garden.Registration.AutoConnect', false);
+
+            // Let's try to find a matching user.
+            if ($autoConnect) {
+                $user = $this->getUserByEmail($ssoInfo);
+
+                // Make sure that the user isn't already linked to another ID.
+                if ($user) {
+                    $result = $this->userModel->getAuthenticationByUser($user['UserID'], $ssoInfo['authenticatorID']);
+                    if ($result) {
+                        // TODO: We should probably add some sort of warning about this.
+                        $user = false;
+                    }
+                }
+            }
+
+            // Try to create a new user since none are matching.
+            if (!$user) {
+                $user = $this->createUser($ssoInfo);
+            }
+
+            // Yay!
+            if ($user !== false) {
+                $this->userModel->saveAuthentication([
+                    'UserID' => $user['UserID'],
+                    'Provider' => $ssoInfo['authenticatorID'],
+                    'UniqueID' => $ssoInfo['uniqueID']
+                ]);
+            }
+        }
+
+        if ($user) {
+            $this->session->start($user['UserID']);
+
+            if ($ssoInfo['authenticatorIsTrusted']) {
+                // Synchronize user's info.
+                $syncInfo = $this->config->get('Garden.Registration.ConnectSynchronize', true);
+
+                // Synchronize user's roles only on registration.
+                $syncRolesOnlyRegistration = $this->config->get('Garden.SSO.SyncRolesOnRegistrationOnly', false);
+
+                // This coupling sucks but I feel like that's the best way to accommodate the config!
+                if ($syncRolesOnlyRegistration && val('connectOption', $ssoInfo) !== 'createuser') {
+                    $syncRoles = false;
+                } else {
+                    // Synchronize user's roles.
+                    $syncRoles = $this->config->get('Garden.SSO.SyncRoles', false);
+                }
+
+                if (!$this->syncUser($ssoInfo, $user, $syncInfo, $syncRoles)) {
```

Comment:
Syncing user info with a trusted provider does not appear to sync roles. Using Vanilla Connect and this branch, `$syncRoles` is always false, even though the connection is trusted.

Reply:
Is Garden.SSO.SyncRoles  set to true?

Reply:
Fixed